### PR TITLE
Added method to search nested containers for a specified component

### PR DESCRIPTION
**Pull request by chromiumboy**
Adds two new methods to SharedContainerSystem that allow you to recursively search the parent objects of a contained entity for a specified component

The first method returns the first matching component found, and the second returns a list of matching components found across all recursive parent entities

### CODE REVIEW

File: Robust.Shared/Containers/SharedContainerSystem.cs [hunk 1]
```diff
@@ -220,6 +220,83 @@ public bool IsEntityInContainer(EntityUid uid, MetaDataComponent? meta = null)
...
+        /// <summary>
+        ///     Finds the first instance of a component on the recursive parented containers that hold an entity  
+        /// </summary>
+        public bool TryFindComponentOnEntityContainerOrParent<T>(
+            EntityUid uid,
+            out T? foundComponent,
+            MetaDataComponent? meta = null,
+            TransformComponent? xform = null,
+            EntityQuery<MetaDataComponent>? metas = null,
+            EntityQuery<TransformComponent>? xforms = null) where T : Component
```
Comment (metalgearsloth) [inline Robust.Shared/Containers/SharedContainerSystem.cs]: Just put the entityquery on the system and also require the user to pass in EntityQuery<T>

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Robust.Shared/Containers/SharedContainerSystem.cs [hunk 2]
```diff
@@ -220,6 +226,70 @@ public bool IsEntityInContainer(EntityUid uid, MetaDataComponent? meta = null)
...
+                return foundComponents.Any();
+
+            if (entityQuery.Resolve(xform.ParentUid, ref foundComponent, false) && foundComponent != null)
+                foundComponents.Add(foundComponent);
+
+            if (TryFindComponentsOnEntityContainerOrParent(xform.ParentUid, entityQuery, out var extraComponents))
+                foundComponents = foundComponents.Concat(extraComponents).ToList();
+
+            return foundComponents.Any();
+        }
```
Comment (metalgearsloth) [inline Robust.Shared/Containers/SharedContainerSystem.cs]: This is going to allocate a new list for every parent which is expensive, better to just pass it in and re-use it. This also avoids the additional Concat and ToList calls.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Robust.Shared/Containers/SharedContainerSystem.cs (hunk 2 above)
Comment (chromiumboy) [reply Robust.Shared/Containers/SharedContainerSystem.cs]: Fair point, the requested changes have been made

File: Robust.Shared/Containers/SharedContainerSystem.cs [hunk 3]
```diff
@@ -220,6 +226,64 @@ public bool IsEntityInContainer(EntityUid uid, MetaDataComponent? meta = null)
...
+                return foundComponents.Any();
+
+            if (!xform.ParentUid.Valid)
+                return foundComponents.Any();
+
+            if (entityQuery.Resolve(xform.ParentUid, ref foundComponent, false) && foundComponent != null)
+                foundComponents.Add(foundComponent);
+
+            return TryFindComponentsOnEntityContainerOrParent(xform.ParentUid, entityQuery, ref foundComponents);
+        }
```
Comment (metalgearsloth) [inline Robust.Shared/Containers/SharedContainerSystem.cs]: You're setting the list to null every loop when it should just be passed in.

File: Robust.Shared/Containers/SharedContainerSystem.cs [hunk 4]
```diff
@@ -220,6 +226,64 @@ public bool IsEntityInContainer(EntityUid uid, MetaDataComponent? meta = null)
...
+        ///     Finds all instances of a component on the recursive parented containers that hold an entity  
+        /// </summary>
+        public bool TryFindComponentsOnEntityContainerOrParent<T>(
+            EntityUid uid,
+            EntityQuery<T> entityQuery,
+            ref List<T> foundComponents,
+            MetaDataComponent? meta = null,
+            TransformComponent? xform = null) where T : Component
+        {
+            T? foundComponent = null;
```
Comment (metalgearsloth) [inline Robust.Shared/Containers/SharedContainerSystem.cs]: Remove this as it looks like you don't even use it.

File: Robust.Shared/Containers/SharedContainerSystem.cs [hunk 5]
```diff
@@ -220,6 +226,64 @@ public bool IsEntityInContainer(EntityUid uid, MetaDataComponent? meta = null)
...
+            return TryFindComponentOnEntityContainerOrParent(xform.ParentUid, entityQuery, ref foundComponent);
+        }
+
+        /// <summary>
+        ///     Finds all instances of a component on the recursive parented containers that hold an entity  
+        /// </summary>
+        public bool TryFindComponentsOnEntityContainerOrParent<T>(
+            EntityUid uid,
+            EntityQuery<T> entityQuery,
+            ref List<T> foundComponents,
```
Comment (metalgearsloth) [inline Robust.Shared/Containers/SharedContainerSystem.cs]: Don't need to pass ref types by ref.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Robust.Shared/Containers/SharedContainerSystem.cs (hunk 4 above)
Comment (chromiumboy) [reply Robust.Shared/Containers/SharedContainerSystem.cs]: foundComponent is required for the entity query on line 281 and 282

File: Robust.Shared/Containers/SharedContainerSystem.cs (hunk 5 above)
Comment (chromiumboy) [reply Robust.Shared/Containers/SharedContainerSystem.cs]: Not sure what you mean here, don't we need to pass the ref for the recursion to work on line 284?

File: Robust.Shared/Containers/SharedContainerSystem.cs (hunk 4 above)
Comment (metalgearsloth) [reply Robust.Shared/Containers/SharedContainerSystem.cs]: Just use TryGetComponent instead of a dummy resolve.

**Pull request merged**